### PR TITLE
Initial connexion implementation

### PR DESCRIPTION
**Pull request by fepas**
[DONE] Adding the connexion library and configuring cachito-api to use connexion in app.py
[DONE] Routing the API endpoints to their respective cachito-api functions 
[**WIP**] Updating unit tests as necessary

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [ ] New code has type annotations
- [x] OpenAPI schema is updated (if applicable)
- [ ] ~DB schema change has corresponding DB migration (if applicable)~
- [ ] ~README updated (if worker configuration changed, or if applicable)~

### CODE REVIEW

**Comment by fepas** [comment]
I'm stuck on the part about fixing the broken unit tests.

image (1)

I started trying to fix the test:  `tests/test_api_v1.py::test_create_and_fetch_request`, 

 I got [this](https://pastebin.com/sEVw1Fb5) result (getting a 404 from the client).

I tried adding the following fixture to _conftest.py_, using [this stackoverflow answer](https://stackoverflow.com/questions/42934525/how-to-test-a-connexion-flask-app) as a base (with the necessary changes related to our code), but I kept getting the same error.

```python
flask_app = connexion.FlaskApp(__name__)
flask_app.add_api('swagger.yml')


@pytest.fixture(scope='module')
def client():
    with flask_app.app.test_client() as c:
        yield c

```

Any suggestion? :D

File: cachito/web/static/api_v1.yaml [hunk 1]
```diff
@@ -228,8 +232,6 @@ paths:
           application/json:
             schema:
               $ref: '#/components/schemas/RequestNew'
-      security:
```
Comment (MartinBasti) [inline cachito/web/static/api_v1.yaml:231]: Why was this removed?

Comment (fepas) [reply cachito/web/static/api_v1.yaml:231]: Forgot to mention this: we get this error when security is present:

    ... Unsupported http authorization scheme negotiate

Talking to @taylormadore yesterday, we started thinking that maybe connexion probably doesn't need to know about or care about kerberos (kerberos isn't even supported by OpenAPI). What do you think @MartinBasti?

Comment (brunoapimentel) [reply cachito/web/static/api_v1.yaml:231]: I got this test running locally changing this:
```
diff --git a/cachito/web/api_v1.py b/cachito/web/api_v1.py
index d840d86..0c51b55 100644
--- a/cachito/web/api_v1.py
+++ b/cachito/web/api_v1.py
@@ -153,16 +153,16 @@ def get_requests():
     return flask.jsonify(response)
 
 
-def get_request(request_id):
+def get_request(id):
     """
     Retrieve details for the given request.
 
-    :param int request_id: the value of the request ID
+    :param int id: the value of the request ID
     :return: a Flask JSON response
     :rtype: flask.Response
     :raise NotFound: if the request is not found
     """
-    json = Request.query.get_or_404(request_id).to_json()
+    json = Request.query.get_or_404(id).to_json()
 
     if json["state"] == RequestStateMapping.complete.name:
         package_count = len(json["packages"])
@@ -171,7 +171,7 @@ def get_request(request_id):
         flask.current_app.logger.info(
             "Returning data for request %i. Found %i packages and %i dependencies. "
             "The following package managers were used: %s.",
-            request_id,
+            id,
             package_count,
             dependency_count,
             json["pkg_managers"],
diff --git a/cachito/web/models.py b/cachito/web/models.py
index 8c0523a..7058608 100644
--- a/cachito/web/models.py
+++ b/cachito/web/models.py
@@ -439,13 +439,13 @@ class Request(db.Model):
 
         if verbose:
             rv["configuration_files"] = flask.url_for(
-                "api_v1.get_request_config_files", request_id=self.id, _external=True
+                "/api/v1.cachito_web_api_v1_get_request_config_files", id=self.id, _external=True
             )
             rv["content_manifest"] = flask.url_for(
-                "api_v1.get_request_content_manifest", request_id=self.id, _external=True
+                "/api/v1.cachito_web_api_v1_get_request_content_manifest", id=self.id, _external=True
             )
             rv["environment_variables_info"] = flask.url_for(
-                "api_v1.get_request_environment_variables", request_id=self.id, _external=True
+                "/api/v1.cachito_web_api_v1_get_request_environment_variables", id=self.id, _external=True
             )
             # Use this list comprehension instead of a RequestState.to_json method to avoid
             # including redundant information about the request itself
@@ -469,7 +469,7 @@ class Request(db.Model):
             if flask.current_app.config["CACHITO_REQUEST_FILE_LOGS_DIR"]:
                 rv["logs"] = {
                     "url": flask.url_for(
-                        "api_v1.get_request_logs", request_id=self.id, _external=True
+                        "/api/v1.cachito_web_api_v1_get_request_logs", id=self.id, _external=True
                     )
                 }
         else:
```

Comment (MartinBasti) [reply cachito/web/static/api_v1.yaml:231]: I think that such change should be mentioned in commit message with explanation why

Comment (MartinBasti) [reply cachito/web/static/api_v1.yaml:231]: @brunoapimentel can you pls explain me, how your diff is related to krb negotiate auth? I miss this magic

Comment (brunoapimentel) [reply cachito/web/static/api_v1.yaml:231]: I posted it to the wrong comment :sweat_smile:. (Deleted it to avoid further confusion)

**Comment by brunoapimentel** [comment]
> I'm stuck on the part about fixing the broken unit tests.
> 
> image (1)
> 
> I started trying to fix the test: `tests/test_api_v1.py::test_create_and_fetch_request`,
> 
> I got [this](https://pastebin.com/sEVw1Fb5) result (getting a 404 from the client).
> 
> I tried adding the following fixture to _conftest.py_, using [this stackoverflow answer](https://stackoverflow.com/questions/42934525/how-to-test-a-connexion-flask-app) as a base (with the necessary changes related to our code), but I kept getting the same error.
> 
> ```python
> flask_app = connexion.FlaskApp(__name__)
> flask_app.add_api('swagger.yml')
> 
> 
> @pytest.fixture(scope='module')
> def client():
>     with flask_app.app.test_client() as c:
>         yield c
> ```
> 
> Any suggestion? :D

The following tests made the test pass locally:

```
diff --git a/cachito/web/api_v1.py b/cachito/web/api_v1.py
index d840d86..0c51b55 100644
--- a/cachito/web/api_v1.py
+++ b/cachito/web/api_v1.py
@@ -153,16 +153,16 @@ def get_requests():
     return flask.jsonify(response)
 
 
-def get_request(request_id):
+def get_request(id):
     """
     Retrieve details for the given request.
 
-    :param int request_id: the value of the request ID
+    :param int id: the value of the request ID
     :return: a Flask JSON response
     :rtype: flask.Response
     :raise NotFound: if the request is not found
     """
-    json = Request.query.get_or_404(request_id).to_json()
+    json = Request.query.get_or_404(id).to_json()
 
     if json["state"] == RequestStateMapping.complete.name:
         package_count = len(json["packages"])
@@ -171,7 +171,7 @@ def get_request(request_id):
         flask.current_app.logger.info(
             "Returning data for request %i. Found %i packages and %i dependencies. "
             "The following package managers were used: %s.",
-            request_id,
+            id,
             package_count,
             dependency_count,
             json["pkg_managers"],
diff --git a/cachito/web/models.py b/cachito/web/models.py
index 8c0523a..7058608 100644
--- a/cachito/web/models.py
+++ b/cachito/web/models.py
@@ -439,13 +439,13 @@ class Request(db.Model):
 
         if verbose:
             rv["configuration_files"] = flask.url_for(
-                "api_v1.get_request_config_files", request_id=self.id, _external=True
+                "/api/v1.cachito_web_api_v1_get_request_config_files", id=self.id, _external=True
             )
             rv["content_manifest"] = flask.url_for(
-                "api_v1.get_request_content_manifest", request_id=self.id, _external=True
+                "/api/v1.cachito_web_api_v1_get_request_content_manifest", id=self.id, _external=True
             )
             rv["environment_variables_info"] = flask.url_for(
-                "api_v1.get_request_environment_variables", request_id=self.id, _external=True
+                "/api/v1.cachito_web_api_v1_get_request_environment_variables", id=self.id, _external=True
             )
             # Use this list comprehension instead of a RequestState.to_json method to avoid
             # including redundant information about the request itself
@@ -469,7 +469,7 @@ class Request(db.Model):
             if flask.current_app.config["CACHITO_REQUEST_FILE_LOGS_DIR"]:
                 rv["logs"] = {
                     "url": flask.url_for(
-                        "api_v1.get_request_logs", request_id=self.id, _external=True
+                        "/api/v1.cachito_web_api_v1_get_request_logs", id=self.id, _external=True
                     )
                 }
         else:
 ```

File: cachito/web/api_v1.py [hunk 2]
```diff
@@ -156,7 +153,6 @@ def get_requests():
     return flask.jsonify(response)
 
 
-@api_v1.route("/requests/<int:request_id>", methods=["GET"])
 def get_request(request_id):
```
Comment (chmeliik) [inline cachito/web/api_v1.py:156]: Still need to rename the parameter names to match what they are called in the OpenAPI spec, IIRC

Comment (chmeliik) [reply cachito/web/api_v1.py:156]: Ah nevermind, I guess it was only the url that needed updating

            rv["configuration_files"] = flask.url_for(
                "/api/v1.cachito_web_api_v1_get_request_config_files",
                request_id=self.id,
                _external=True,

File: requirements-web.in [hunk 3]
```diff
@@ -6,4 +7,5 @@ greenlet>=1.1.0
 psycopg2-binary
 pydantic
 SQLAlchemy<1.5
+swagger-ui-bundle
```
[the remark this answers is not in the document]
Comment (fepas) [reply requirements-web.in:10]: Connexion is installed with:

    pip install connexion[swagger-ui]

So we need it so that there are no problems with the swagger. 
(Or I think we can also disable the swagger, but I find it quite useful, to be honest.)

[the remark this answers is not in the document]
Comment (chmeliik) [reply requirements-web.in:10]: Oh, so is this like a replacement for [the API docs](https://containerbuildsystem.github.io/cachito/)? Or something different? https://github.com/spec-first/connexion#the-swagger-ui-console

Maybe this deserves a separate PR with a docs update?

[the remark this answers is not in the document]
Comment (taylormadore) [reply requirements-web.in:10]: Yeah, connexion includes the swagger-ui by default (it can also be disabled). It's pretty similar to the API docs we currently have with redoc, so we should probably pick one or the other.

**Comment by lkolacek** [comment]
Can we squash commits here, please?

File: requirements-web.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply requirements-web.in:10]: Should we disable it for now and, if we want it, do it in a follow-up PR/story?

[the remark this answers is not in the document]
Comment (MartinBasti) [reply requirements-web.in:10]: please don't remove swagger :)

[the remark this answers is not in the document]
Comment (chmeliik) [reply requirements-web.in:10]: So we keep both documentations? http://localhost:8080 and http://localhost:8080/api/v1/ui ?

Maybe with a follow-up story to replace the root page with http://localhost:8080/api/v1/ui, or at least document that the latter one exists?

File: tests/test_api_v1.py [hunk 4]
```diff
@@ -566,10 +566,10 @@ def test_fetch_paginated_requests(
     assert fetched_requests[0]["dependencies"] == len(sample_deps_replace)
     assert fetched_requests[0]["packages"] == 1
 
-    # Invalid per_page defaults to 10
+    # Invalid per_page returns error
     rv = client.get("/api/v1/requests?per_page=tom_hanks")
-    assert len(rv.json["items"]) == 10
-    assert response["meta"]["per_page"] == 10
+    assert rv.json["detail"] == "Wrong type, expected 'integer' for query parameter 'per_page'"
+    assert rv.json["status"] == 400
```
Comment (chmeliik) [inline tests/test_api_v1.py:568]: technically a breaking change, but should be OK

File: tests/test_api_v1.py [hunk 5]
```diff
@@ -2458,34 +2458,15 @@ def test_get_content_manifests_by_requests(app, client, db, auth_env, tmpdir):
...
         ],
-        [
-            "pkg_manager=",
-            3,
-            [
-                "https://github.com/org/bar.git",
-                "https://github.com/org/baz.git",
-                "https://github.com/org/foo.git",
-            ],
-        ],
```
Comment (chmeliik) [inline tests/test_api_v1.py:2445]: So I take it we lost the ability to specify empty params? But this seems like one of the more OK breaking changes

File: tests/test_api_v1.py [hunk 6]
```diff
@@ -1707,7 +1706,11 @@ def test_request_patch_invalid(
 
     rv = client.patch(f"/api/v1/requests/{request_id}", json=payload, environ_base=worker_auth_env)
     assert rv.status_code == status_code
-    assert rv.json == {"error": message}
+
+    if "detail" in rv.json:
+        assert message in rv.json["detail"]
+    else:
+        assert message in rv.json["error"]
```
Comment (chmeliik) [inline tests/test_api_v1.py:1713]: Did half of our error messages move to "detail" while the other half stayed in "error"? That might be on the not-OK side of breaking changes

File: tests/test_api_v1.py [hunk 7]
```diff
@@ -1112,8 +1110,17 @@ def test_create_request_invalid_packages(packages, pkg_managers, error_msg, auth
...
     rv = client.post("/api/v1/requests", json=data, environ_base=auth_env)
+
     assert rv.status_code == 400
-    assert re.match(error_msg, rv.json["error"])
+
+    if "error" in rv.json:
+        if list(packages.values())[0] == [{}]:
+            assert "must be an array of objects" in rv.json["error"]
+        else:
+            assert re.match(error_msg, rv.json["error"])
```
Comment (chmeliik) [inline tests/test_api_v1.py:1120]: why this logic rather than having the right error message in test params?

File: tests/test_api_v1.py (hunk 6 above)
Comment (chmeliik) [reply tests/test_api_v1.py:1713]: ^ would break OSBS https://github.com/containerbuildsystem/atomic-reactor/blob/10dce319b239a2bb93cdd78c4506fb2fa7103b80/atomic_reactor/utils/cachito.py#L97

File: requirements-web.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (MartinBasti) [reply requirements-web.in:10]: well, keep it at current location, and for any other changes, open a new story

File: tests/test_api_v1.py (hunk 6 above)
Comment (taylormadore) [reply tests/test_api_v1.py:1713]: I remember finding a way to override the format of connexion's error response and return ValidationError just like we do today. :thinking:

Comment (taylormadore) [reply tests/test_api_v1.py:1713]: I think the solution was in creating a [custom validator](https://connexion.readthedocs.io/en/latest/request.html#custom-validators)

File: cachito/web/app.py [hunk 8]
```diff
@@ -110,6 +115,14 @@ def create_app(config_obj=None):
 
     app.register_blueprint(docs)
     app.register_blueprint(api_v1, url_prefix="/api/v1")
+
+    path = Path(__file__).parents[1].absolute()
```
Comment (brunoapimentel) [inline cachito/web/app.py:119]: We could change this to `Path(__file__).parent.absolute()` and then reference the file with `f"{path}/static/api_v1.yaml"`

File: cachito/web/validation.py [hunk 9]
```diff
@@ -46,3 +49,55 @@ def validate_dependency_replacements(replacements: List[Dict[str, Any]]) -> None
         raise ValidationError('"dependency_replacements" must be an array')
     for replacement in replacements:
         validate_replacement(replacement)
+
+
+class RequestBodyValidator(decorators.validation.RequestBodyValidator):
+    """
+    This class overrides the default connexion RequestBodyValidator.
+
+    It returns the complete string representation of the
```
Comment (brunoapimentel) [inline cachito/web/validation.py:58]: Doesn't the original exception also return the complete error message? 

I thought this change was to keep the format and exception type we expect in Cachito (ValidationError, with `{"error": "message"}` instead of `{"detail": "message"}`). 

Correct if I`m wrong, though :sweat_smile:.

File: cachito/web/validation.py [hunk 10]
```diff
@@ -46,3 +49,55 @@ def validate_dependency_replacements(replacements: List[Dict[str, Any]]) -> None
...
+        - https://connexion.readthedocs.io/en/latest/request.html
+    """
+
+    def validate_schema(self, data, url):
+        """Throw cachito.ValidationError."""
+        if self.is_null_value_valid and jsonschema.is_null(data):
+            return None
+        try:
+            self.validator.validate(data)
+        except Exception as exception:
```
Comment (brunoapimentel) [inline cachito/web/validation.py:72]: The original function catches only `jsonschema.ValidationError` error types. Shouldn't we do the same here? This way, we could potentially hide other error types.

File: cachito/web/validation.py [hunk 11]
```diff
@@ -46,3 +49,55 @@ def validate_dependency_replacements(replacements: List[Dict[str, Any]]) -> None
...
+
+    def __call__(self, function):
+        """Throw cachito.ValidationError."""
+        wrapper = super().__call__(function)
+
+        def handle_wrapper(request):
+            """Handle original wrapper."""
+            try:
+                return wrapper(request)
+            except Exception as e:
```
Comment (brunoapimentel) [inline cachito/web/validation.py:98]: I think it would be clearer here if we caught only the specific Connexion exceptions here (ExtraParameterProblem or BadRequestProblem) instead of checking for existence of a `details` attribute.

Any non-Connexion exception would be re-raised, as it is currently done here.

**Comment by brunoapimentel** [comment]
I personally don't see much value into separating each fix for the unit tests into a single commit. I'd squash all FIX commits here.

But that's me, let's see what others think.

File: cachito/web/validation.py (hunk 9 above)
Comment (chmeliik) [reply cachito/web/validation.py:58]: Yeah, IIRC the only change we needed was to have "error" in the JSON returned to the user (not "detail")

File: cachito/web/validation.py (hunk 11 above)
Comment (brunoapimentel) [reply cachito/web/validation.py]: If I make a request such as `api/v1/requests/2?hello=1`, I still get:
```
{
"detail": "Extra query parameter(s) hello not in spec",
"status": 400,
"title": null,
"type": "about:blank"
}
```
There's no unit test that covers extra URL query params, maybe we should add one.

**Review by chmeliik — APPROVED** [review]
LGTM :tada: 

I see that we're merging this into a "feature branch" - https://github.com/containerbuildsystem/cachito/tree/CLOUDBLD-5466. +1 for that, seems safer to thoroughly this before merging to main.

File: cachito/web/app.py [hunk 12]
```diff
@@ -110,6 +115,14 @@ def create_app(config_obj=None):
 
     app.register_blueprint(docs)
     app.register_blueprint(api_v1, url_prefix="/api/v1")
```
Comment (taylormadore) [inline cachito/web/app.py:117]: I wonder if this doing anything now that we've removed all of the routes that referenced it?

**Comment by taylormadore** [comment]
Do we still need this? I believe we removed all of the other references to it. https://github.com/containerbuildsystem/cachito/blob/f626d97ea669e12879a72762e03aa92d862d66a6/cachito/web/static/api_v1.yaml#L1270-L1273

File: tests/test_api_v1.py [hunk 13]
```diff
@@ -2486,8 +2476,9 @@ def test_filter_requests(
 
     if expected_items_count is None:
         assert HTTPStatus.BAD_REQUEST == rv.status_code
-        assert expected_repos in rv.data.decode()
+        assert expected_repos in rv.json["error"]
     else:
+        print("!", rv.json)
```
Comment (brunoapimentel) [inline tests/test_api_v1.py:2481]: Should this print be here?

File: cachito/web/validation.py [hunk 14]
```diff
@@ -46,3 +50,47 @@ def validate_dependency_replacements(replacements: List[Dict[str, Any]]) -> None
         raise ValidationError('"dependency_replacements" must be an array')
     for replacement in replacements:
         validate_replacement(replacement)
+
+
+class RequestBodyValidator(decorators.validation.RequestBodyValidator):
```
Comment (brunoapimentel) [inline cachito/web/validation.py:55]: These classes do not have type annotations... although the base classes also don't, so it's probably fine.

File: tests/test_api_v1.py (hunk 13 above)
Comment (fepas) [reply tests/test_api_v1.py]: No. xD
